breaker.go

```diff
@@ -0,0 +1,228 @@
+package hoglet
+
+import (
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// untypedCircuit is used to avoid type annotations when implementing a breaker.
+type untypedCircuit interface {
+	stateForCall() State
+	setOpenedAt(int64)
+}
+
+type observer interface {
+	// observe is called after the wrapped function returns. If [Circuit.Do] returns a non-nil [Observable], it will be
+	// called exactly once.
+	observe(failure bool)
+}
+
+// newObservableCall creates a new [Observable] that ensures it can only be observed a single time.
+func newObservableCall(f func(bool)) observer {
+	o := sync.Once{}
+	return observableCall(func(failure bool) {
+		o.Do(func() {
+			f(failure)
+		})
+	})
+}
+
+// observableCall tracks a single call through the breaker.
+// It should be instantiated via [newObservableCall] to ensure the observer is only called once.
+type observableCall func(bool)
+
+func (o observableCall) observe(failure bool) {
+	o(failure)
+}
+
+// EWMABreaker is a [Breaker] that uses an exponentially weighted moving failure rate. See [NewEWMABreaker] for details.
+//
+// A zero EWMABreaker is ready to use, but will never open.
+type EWMABreaker struct {
+	decay     float64
+	threshold float64
+	circuit   untypedCircuit
+
+	// State
+	failureRate atomic.Value
+}
+
+// NewEWMABreaker creates a new [EWMABreaker] with the given sample count and threshold. It uses an Exponentially
+// Weighted Moving Average to calculate the current failure rate.
+//
+// ⚠️ This is an observation-based breaker, which means it requires new calls to be able to update the failure rate, and
+// therefore REQUIRES the circuit to set a half-open threshold via [WithHalfOpenDelay]. Otherwise an open circuit will
+// never observe any successes and thus never close.
+//
+// Compared to the [SlidingWindowBreaker], this breaker responds faster to failure bursts, but is more lenient with
+// constant failure rates.
+//
+// The sample count is used to determine how fast previous observations "decay". A value of 1 causes a single sample to
+// be considered. A higher value slows down convergence. As a rule of thumb, breakers with higher throughput should use
+// higher sample counts to avoid opening up on small hiccups.
+//
+// The failureThreshold is the failure rate above which the breaker should open (0.0-1.0).
+func NewEWMABreaker(sampleCount int, failureThreshold float64) *EWMABreaker {
```

Comment:
Maybe we should validate, that failureThreshold is really between 0 and 1

Reply:
I'm a bit unsure about this: it's a bit safer, but it makes the API considerably less ergonomic (then you'd have to create the breaker first, check for errors, and then create the circuit). Since these values are usually not dynamic, maybe we can  live with it? 

Another option would be turning the breakers into normal `Option`s and erroring out on `NewCircuit` if no breaker is set.